shelf_router/CHANGELOG.md

```diff
@@ -1,3 +1,11 @@
+## v0.8.0-nullsafety.0
+
+ * Migrate package to null-safety
+ * Since handlers are not allowed to return `null` in `shelf` 1.0.0, a router
+   will return a default 404 response instead.
+   This behavior can be overridden with the `notFoundHandler` constructor
+   parameter.
```

Comment:
IMHO making handlers return a non-nullable response wasn't a great call, but that's what it is...

`Cascade` allows providing a custom callback to check whether a response is acceptable or not. We could have something like that in the router as well.

Reply:
> `Cascade` allows providing a custom callback to check whether a response is acceptable or not. We could have something like that in the router as well.

This seems very inconvenient and error-prone to use.. Sure you could create your response instances with a header or something you can use to check if they should be accepted or not. But it feels fragile. 


> IMHO making handlers return a non-nullable response wasn't a great call, but that's what it is...

Yeah, either allow returning null, or have a sentinel value in `package:shelf` to indicate no response from a handler would be useful.

But introducing a sentinel value to `Cascade` is probably rather breaking by now. Though `Cascade` feels a bit useless when you realize that a handler can't forcibly return `404`, at-least not without fragile hacks.

@kevmoo, did we consider these use-cases when we made `Handler` require a non-null response?

-----------------------

I suppose requiring a response is quite useful in general, and for `package:shelf_router` reasonably acceptable, because you rarely need the cascade effect anyways. It did make it easy to build middleware, but that might be better handled by introducing a `router.use(String route, shelf.Middleware middleware)` method (if that can be made to work).